pkg/util/util_test.go

```diff
@@ -0,0 +1,246 @@
+			clusterCidr := []string{"1.2.3.4/16"}
+			serviceCidr := []string{"5.6.7.8/16"}
+			colorCodes := []string{"red"}
+			cluster, err := GetLocalCluster(types.SubmarinerSpecification {
+				ClusterId: clusterId,
+				ClusterCidr: clusterCidr,
+				ServiceCidr: serviceCidr,
+				ColorCodes: colorCodes,
+			})
+
+			Expect(err).ToNot(HaveOccurred())
+			Expect(cluster.ID).To(Equal(clusterId))
+			Expect(cluster.Spec.ClusterID).To(Equal(clusterId))
+			Expect(cluster.Spec.ServiceCIDR).To(Equal(serviceCidr))
+			Expect(cluster.Spec.ClusterCIDR).To(Equal(clusterCidr))
+			Expect(cluster.Spec.ColorCodes).To(Equal(colorCodes))
+		})
+	})
+
+	Describe("Function GetLocalEndpoint", func() {
+		It("should return a valid SubmarinerEndpoint object", func() {
+			subnets := []string{"1.2.3.4/16"}
+			endpoint, err := GetLocalEndpoint("east", "backend", map[string]string{}, false, subnets)
+
+			Expect(err).ToNot(HaveOccurred())
+			Expect(endpoint.Spec.ClusterID).To(Equal("east"))
+			Expect(endpoint.Spec.CableName).To(HavePrefix("submariner-cable-east-"))
+			Expect(endpoint.Spec.Hostname).NotTo(Equal(""))
+			Expect(endpoint.Spec.PrivateIP).NotTo(BeNil())
+			Expect(endpoint.Spec.Backend).To(Equal("backend"))
+			Expect(endpoint.Spec.Subnets).To(Equal(subnets))
+			Expect(endpoint.Spec.NATEnabled).To(Equal(false))
+		})
+	})
+
+	Describe("Function GetClusterIdFromCableName", func() {
+		Context("With a simple embedded cluster ID", func() {
+			It("should extract and return the cluster ID", func() {
+				Expect(GetClusterIdFromCableName("submariner-cable-east-172-16-32-5")).To(Equal("east"))
+			})
+		})
+	
+		Context("With an embedded cluster ID containing dashes", func() {
+			It("should extract and return the cluster ID", func() {
+				Expect(GetClusterIdFromCableName("submariner-cable-my-super-long_cluster-id-172-16-32-5")).To(
+					Equal("my-super-long_cluster-id"))
+			})
+		})
+	})
+
+	Describe("Function GetEndpointCRDName", func() {
+		Context("With valid SubmarinerEndpoint input", func() {
+			It("should return <cluster ID>-<cable name>", func() {
+				name, err := GetEndpointCRDName(types.SubmarinerEndpoint {
+					Spec: subv1.EndpointSpec{
+						ClusterID: "ClusterID",
+						CableName: "CableName",
+					},
+				})
+
+				Expect(err).ToNot(HaveOccurred())
+				Expect(name).To(Equal("ClusterID-CableName"))
+			})
+		})
+
+		Context("With a nil cluster ID", func() {
+			It("should return an error", func() {
+				_, err := GetEndpointCRDName(types.SubmarinerEndpoint {
+					Spec: subv1.EndpointSpec{
+						CableName: "CableName",
+					},
+				})
+
+				Expect(err).To(HaveOccurred())
+			})
+		})
+		
+		Context("With a nil cable name", func() {
+			It("should return an error", func() {
+				_, err := GetEndpointCRDName(types.SubmarinerEndpoint {
+					Spec: subv1.EndpointSpec{
+						ClusterID: "ClusterID",
+					},
+				})
+
+				Expect(err).To(HaveOccurred())
+			})
+		})
+	})
+
+	Describe("Function GetClusterCRDName", func() {
```

Comment:
(No action required) `GetClusterCRDName` seems like another case where inlining would be appropriate.  The 2 call sites could as well check the length of `Spec.ClusterID` instead of checking an error returned by a different function.  

Also, kube types rely on admission control to ensure the validity of api calls.  Maybe it would make sense for submariner to validate its internal types on creation and modification rather than spreading knowledge about the type invariants types throughout the codebase?

Reply:
maybe. But beyond the scope of this patch.